### PR TITLE
Add haskell language support to pre-commit.

### PR DESCRIPTION
**Pull request by alunduil**
I'm attempting to follow the other languages and documentation in CONTRIBUTING (which seems outdated) to add Haskell support to pre-commit so I can add hooks for several useful Haskell tools.

I'm running into issues when trying to run tests of my new language, and I'm trying to figure out where I'm going wrong. Can someone provide guidance?

I'm assuming the system has Haskell installed already, the same as Python and attempting to follow the documented interface in CONTRIBUTING as well as I can without adding too many things for a first pass. I'm hitting strange errors where my mock project can be listed, but it seems to be something other than the CWD for the installation (no project file found to do the install of '.').

Please let me know if there is more up-to-date documentation or something I should try. Once I pass the tests, I'll be ready for deeper integration with some tools to see how things go.

### CODE REVIEW

File: pre_commit/languages/haskell.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                additional_dependencies,
+                into=environment_dir,
+            ),
+            check=True,
+        )
+
+
+def _cabal_install_command(
+    packages: typing.Iterable[str],
+    into: pathlib.Path | None = None,
```
Comment (asottile) [inline pre_commit/languages/haskell.py:69]: this should not be optional -- it should *always* be installing into a target directory

File: pre_commit/languages/haskell.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    _LOGGER.debug(f'environment_dir: {environment_dir}')
+    print(f'pathlib.Path.cwd(): {pathlib.Path.cwd()}')
+    _LOGGER.debug(f'pathlib.Path.cwd(): {pathlib.Path.cwd()}')
+
+    cabal_files = [str(path) for path in pathlib.Path.cwd().glob('*.cabal')]
+    assert len(cabal_files) == 1, (
+        f"didn't find a unique cabal file: {', '.join(cabal_files)}"
+    )
+
+    subprocess.run(
```
Comment (asottile) [inline pre_commit/languages/haskell.py:53]: there shouldn't be any usage of `subprocess` directly -- these should go through `setup_cmd` so they have the proper location -- this is probably most of your issues with getting this to work

File: pre_commit/languages/haskell.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+import contextlib
+import logging
+import pathlib
+import subprocess
+import typing
```
Comment (asottile) [inline pre_commit/languages/haskell.py:7]: names should be from-imported from `typing`

File: tests/languages/haskell_test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            'init',
+            '--non-interactive',
+            '--quiet',
+            '--minimal',
+            '--package-name=example',
+            '--simple',
+        ],
+        cwd=tmp_path,
+        check=True,
+    )
```
[the remark this answers is not in the document]
Comment (alunduil) [reply tests/languages/haskell_test.py:29]: Yep, that was my next planned step due to the differences between CI and my local runs.

File: pre_commit/languages/haskell.py (hunk 2 above)
Comment (alunduil) [reply pre_commit/languages/haskell.py:53]: That sounds like the winner.  I'll give that a shot and see what happens.  Thanks.

**Comment by alunduil** [comment]
@asottile, thanks for the pointers. Let me know if you have any other stylistic or architectural guidance. This is now passing, and I am confident I do what I need to start playing with other repositories to add hooks for this language.

**Comment by asottile** [comment]
I reworked it a little bit, picked a smaller test dependency, and updated it to match the structure of the rest of the languages and language tests

**Review by asottile — APPROVED** [review]

**Pull request merged**